### PR TITLE
feat(s3): switch back to hackney to support vhost-style access

### PR DESCRIPTION
**Pull request by keynslug**
Fixes [EMQX-12940](https://emqx.atlassian.net/browse/EMQX-12940).

Release version: e5.8.3

## Summary

Add an ability to configure _bucket access method_ for applications using `emqx_s3` client, with 2 possible values: _path_ or _vhost_. While not strictly required to support the requested service, also change the underlying HTTP client to `hackney` with its internal pool implementation, to be able to correctly handle [redirects](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTRedirect.html) permitted by S3 APIs.

See individual commits for details.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [x] Schema changes are backward compatible

[EMQX-12940]: https://emqx.atlassian.net/browse/EMQX-12940?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by savonarola** [comment]
Still reading 😅

File: apps/emqx_s3/src/emqx_s3_client.erl [hunk 1]
```diff
@@ -298,186 +302,45 @@ ensure_aws_credentials(
...
-
+%%   HTTP Response --(hackney)-> HackneyHeaders
+%%   HackneyHeaders --(emqx_s3_client_http)--(erlcloud_httpc)-> ErlcloudResponseHeders
+%%   ErlcloudResponseHeders --(emqx_s3_client)-> ...
+%%
 %% UserHeders (emqx_s3 API headers) are maps with string/binary keys.
 %% ErlcloudRequestHeaders are lists of tuples with string keys and iodata values
 %% ErlcloudResponseHeders are lists of tuples with lower case string keys and iodata values.
-%% EhttpcHeaders are lists of tuples with binary keys and iodata values.
+%% HackneyHeaders are lists of tuples with binary keys and iodata values.
```
Comment (savonarola) [inline apps/emqx_s3/src/emqx_s3_client.erl:334]: Should't this module completely forget about header conversion and the `emqx_s3_client_http` actual backend (Hackney)?

File: apps/emqx_s3/src/emqx_s3_client_http.erl [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            ok
+    end.
+
+pool_config(#{
+    transport_options := #{
+        pool_type := _PoolType,
+        pool_size := PoolSize
+    }
+}) ->
+    [
```
Comment (savonarola) [inline apps/emqx_s3/src/emqx_s3_client_http.erl:218]: So hackney pools are not per-host, and we don't need all the "switching pool" mechanics?

Comment (keynslug) [reply apps/emqx_s3/src/emqx_s3_client_http.erl:218]: Right.

Comment (savonarola) [reply apps/emqx_s3/src/emqx_s3_client_http.erl:218]: That's a nice simplification

**Review by savonarola — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-5090)

**Pull request merged**